### PR TITLE
[CI] Add conditional build to macOS pipeline

### PR DESCRIPTION
**Pull request by simon-mo**
## Why are these changes needed?
This PR adds conditional directive for macOS CI jobs. The goal is to reduce end to end CI time by skipping macOS builds that's not affected.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: .buildkite/pipeline.macos.yml [hunk 1]
```diff
@@ -73,17 +75,13 @@ steps:
     # clang-format is needed by java/test.sh
     - pip install clang-format==12.0.1
     - ./java/test.sh
+    - ./ci/ci.sh test_cpp
```
Comment (simon-mo) [inline .buildkite/pipeline.macos.yml:78]: this test takes 5min, doesn't warrant a standalone build

**Comment by simon-mo** [comment]
evidence of empty selection: https://buildkite.com/ray-project/ray-builders-pr/builds/31748#02c32e48-725d-4626-9bbe-7c39438461d2/98-164

**Comment by simon-mo** [comment]
this is the corresponding ci stack commit https://github.com/ray-project/buildkite-ci-stack/commit/454ddaab47e9bcccff8f6b9e0848f05c248d18d1

File: ci/pipeline/determine_tests_to_run.py [hunk 2]
```diff
@@ -102,6 +101,8 @@ def get_commit_range():
 
         # Dry run py_dep_analysis.py to see which tests we would have run.
         try:
+            import py_dep_analysis as pda
```
Comment (simon-mo) [inline ci/pipeline/determine_tests_to_run.py:104]: syntax error otherwise because the script is python3

**Comment by simon-mo** [comment]
This PR is ready to be merged as long as lint passes

**Review by mwtian — APPROVED** [review]

**Pull request merged**